src/Campaigns/Repositories/CampaignRepository.php

```diff
@@ -138,6 +140,50 @@ public function update(Campaign $campaign): void
         Hooks::doAction('givewp_campaign_updated', $campaign);
     }
 
+    /**
+     * @unreleased
+     *
+     * @throws Exception
+     */
+    public function addCampaignForm(Campaign $campaign, DonationForm $donationForm, bool $isDefault = false)
+    {
+        Hooks::doAction('givewp_campaign_form_updating', $campaign, $donationForm, $isDefault);
+
+        DB::query('START TRANSACTION');
+
+        try {
+            // Make sure we'll have only one default form
+            if ($isDefault) {
+                DB::table('give_campaign_forms')
+                    ->where('campaign_id', $campaign->id)
+                    ->update([
+                        'is_default' => false,
+                    ]);
+            }
+
+            $table = DB::prefix('give_campaign_forms');
+            DB::query(
+                DB::prepare("INSERT INTO {$table} (form_id, campaign_id, is_default ) VALUES (%s, %s, %s) ON DUPLICATE KEY UPDATE is_default = %s",
+                    [
+                        $donationForm->id,
+                        $campaign->id,
+                        $isDefault,
+                        $isDefault,
+                    ])
+            );
+        } catch (Exception $exception) {
+            DB::query('ROLLBACK');
+
+            Log::error('Failed updating a campaign form', compact('campaign'));
+
+            throw new $exception('Failed updating a campaign form');
```

Comment:
The error message doesn't reflect the action we are doing here.

Reply:
Resolved in 99a801dbdec928b92d8c4bd9fe1bd086c6b66148